vsintegration/src/FSharp.Editor/LanguageService/FSharpProjectOptionsManager.fs

```diff
@@ -334,93 +334,150 @@ type private FSharpProjectOptionsReactor (workspace: Workspace, settings: Editor
-                                        reply.Reply options
-                                else
-                                    reply.Reply None
-                        with
-                        | _ ->
-                            reply.Reply None
 
                 | FSharpProjectOptionsMessage.ClearOptions(projectId) ->
-                    match cache.TryRemove(projectId) with
-                    | true, (_, _, projectOptions) ->
-                        lastSuccessfulCompilations.TryRemove(projectId) |> ignore
-                        checkerProvider.Checker.ClearCache([projectOptions])
-                    | _ ->
-                        ()
-                    legacyProjectSites.TryRemove(projectId) |> ignore
+                    clearOptions projectId
+
                 | FSharpProjectOptionsMessage.ClearSingleFileOptionsCache(documentId) ->
-                    match singleFileCache.TryRemove(documentId) with
-                    | true, (_, _, projectOptions) ->
-                        lastSuccessfulCompilations.TryRemove(documentId.ProjectId) |> ignore
-                        checkerProvider.Checker.ClearCache([projectOptions])
-                    | _ ->
-                        ()
+                    clearSingleFileOptionsCache documentId
         }
 
     let agent = MailboxProcessor.Start((fun agent -> loop agent), cancellationToken = cancellationTokenSource.Token)
 
     member _.TryGetOptionsByProjectAsync(project, ct) =
-        agent.PostAndAsyncReply(fun reply -> FSharpProjectOptionsMessage.TryGetOptionsByProject(project, reply, ct))
+        if useMailboxProcessor then
+            agent.PostAndAsyncReply(fun reply -> FSharpProjectOptionsMessage.TryGetOptionsByProject(project, reply, ct))
+        else
+            asyncLock (tryGetOptionsByProject(project, ct))
 
     member _.TryGetOptionsByDocumentAsync(document, ct, userOpName) =
-        agent.PostAndAsyncReply(fun reply -> FSharpProjectOptionsMessage.TryGetOptionsByDocument(document, reply, ct, userOpName))
+        if useMailboxProcessor then
+            agent.PostAndAsyncReply(fun reply -> FSharpProjectOptionsMessage.TryGetOptionsByDocument(document, reply, ct, userOpName))
+        else
+            asyncLock (tryGetOptionsByDocument(document, ct, userOpName))
 
     member _.ClearOptionsByProjectId(projectId) =
-        agent.Post(FSharpProjectOptionsMessage.ClearOptions(projectId))
+        if useMailboxProcessor then
+            agent.Post(FSharpProjectOptionsMessage.ClearOptions(projectId))
+        else
+            Async.Start(asyncLock (async { clearOptions projectId }))
 
     member _.ClearSingleFileOptionsCache(documentId) =
-        agent.Post(FSharpProjectOptionsMessage.ClearSingleFileOptionsCache(documentId))
+        if useMailboxProcessor then
+            agent.Post(FSharpProjectOptionsMessage.ClearSingleFileOptionsCache(documentId))
+        else
+            Async.Start(asyncLock (async { clearSingleFileOptionsCache documentId }))
```

Comment:
I have to admit, this is quite clever ... even though it gives me the heejeebeejees.

Reply:
@KevinRansom, from my cursory understanding, even:

* if the semantics are mixed up, if I'm debugging the compiler and know the issue, I can work around easily
* it can't impact the consumers of released FCS

The heejeebeejees (not sure what it means) probably should be speculation on "what if I can't step debug and fix the issue once I know if there were even one" but not actual problem, from what I understand.

I've been lost few times, probably to advanced stuff done with CE, and limited "debugger semantic" of all this from dotnet runtime perspective, but I'd personally know how to work around, if I would hit this issue, or troubleshoot people that do.